docs/guides/pallas_kernels_performance.md

```diff
@@ -0,0 +1,144 @@
+* **Algorithmic Optimizations**: Implement novel algorithms not available in standard libraries, such as flash attention, which combines the benefits of the above points to significantly speed up attention and save memory.
+
+## When Not to Use Pallas (and Notable Exceptions)
+
+Pallas is a specialized tool and not always the right choice. Stick with standard JAX in these scenarios:
+
+* **When XLA is Sufficient**: If profiling shows your code is already efficient, a custom kernel adds complexity for little gain. **Always profile first.**
+* **For Most Purely Compute-Bound Operations**: For large, dense operations like `jnp.matmul`, it's hard to beat XLA's highly optimized library calls.
+    
+    **A key exception**, however, is when the data layout is irregular. In **Mixture-of-Experts (MoE) MLPs**, for example, using standard dense matmuls on ragged tensors (where different experts process different numbers of tokens) leads to significant wasted computation. A custom Pallas kernel can avoid this waste and often outperform the generic XLA path in practice.
+
+* **If Maintainability is a Top Priority**: Pallas kernels are lower-level and can be harder to debug. If the performance gain is marginal, the maintenance overhead may not be worth it.
+
+* **Without a Clear Bottleneck**: Avoid premature optimization. Use Pallas only after profiling has identified a specific, memory-bound bottleneck.
+
+
+### Pallas Kernels in MaxText
+
+MaxText uses custom kernels written in Pallas to achieve high performance on TPUs. These kernels optimize key operations within the model, boosting training and inference speed.
+
+* **Splash Attention**: MaxText can use a Pallas-based implementation of Splash Attention for training language models. This is a memory-efficient attention mechanism that processes calculations in smaller, tiled blocks to avoid memory bottlenecks and accelerate training, especially for long sequences. You can find its implementation in `MaxText/kernels/splash_attention_kernel.py`.
+
+* **Paged and Ragged Attention**: For inference, MaxText uses Paged and Ragged Attention kernels. These are highly efficient for handling batched requests with varying sequence lengths by managing the KV cache in non-contiguous memory "pages" and avoiding padding. This is crucial for high-throughput serving and is used across most models. The implementation can be found in `MaxText/inference/paged_attention.py` and `MaxText/inference/paged_attention_kernel_v2.py`.
+
+* **MegaBlox**: For Mixture-of-Experts (MoE) models such as **Mixtral, Qwen, and Deepseek**, we utilize MegaBlox kernels. These are optimized for the sparse Grouped matrix multiplication (GMM) in MoE layers, enabling efficient routing of tokens to experts. It also support quantization for further performance gains. You can see the implementation in `MaxText/kernels/megablox/gmm.py`.
```

Comment:
It makes the matmul efficient, but routing is separate (not using pallas kernel). https://github.com/AI-Hypercomputer/maxtext/blob/83097ddac2d08397429d79de27099ed4b72a898d/MaxText/layers/moe.py#L288